lib/CodeGen/CGExprScalar.cpp

```diff
@@ -2319,6 +2319,108 @@ static BinOpInfo createBinOpInfoFromIncDec(const UnaryOperator *E,
   return BinOp;
 }
 
+// Get the pointer in a pointer arithmetic expression.
+// *(p + i) -> p
+// *(i + p) -> p
+static Expr *peelOffPointerArithmetic(const BinaryOperator *B) {
+  if (B->isAdditiveOp() && B->getType()->isPointerType()) {
+    if (B->getLHS()->getType()->isPointerType())
+      return B->getLHS();
+    else if (B->getRHS()->getType()->isPointerType())
+      return B->getRHS();
+  }
+  return nullptr;
+}
+
+static Expr *peelOffOuterExpr(Expr *E) {
+  if (auto *ICE = dyn_cast<ImplicitCastExpr>(E))
+    return peelOffOuterExpr(ICE->getSubExpr());
+
+  if (const auto *BO = dyn_cast<BinaryOperator>(E))
+    if (Expr *SubE = peelOffPointerArithmetic(BO))
+      return peelOffOuterExpr(SubE);
+
+  return E;
+}
+
+// Determine whether to emit a dynamic non-null check.
+static void emitDynamicNonNullCheck(CodeGenFunction &CGF, Expr *E) {
+  if (!CGF.CGM.getCodeGenOpts().CheckedCNullPtrChecks)
+    return;
```

Comment:
I would structure this code differently.   It's an optimization to not emit a dynamic null check if E can be proved at compile-time to be non-null.  So I'd try to prove that E does not need a null check and return if the null check can be skipped.  Then I would go into the general case. of adding a check.

Please improve the comments on this method.  We should always explain what non-obvious arguments are.  For example, in this case E is the pointer-typed expression in a pointer-arithmetic expression.  We need to check the value of E is non-null, if E is a checked pointer type.